### DISCUSSION TITLE
Edit mode vs. Browse mode

### DISCUSSION THREAD

**Discussion by Gregory1701** [codefori/vscode-ibmi#1331] [IBM i]
I (as well as at least one of my coworkers) think it would be nice if the open source member tab somehow indicated whether you were in Edit mode or Browse mode.  RDi does this by showing a pencil vs a spyglass.

The problem we're looking to resolve is this.

- Ctl+Alt+P and open a source member which exists in my development library as well as in production
- I opened a version of a source member that is in our production library (let's say by accident)
- Start making the necessary changes, then go to save the source member, compile, etc.
- But then I discover that I cannot save it. I do get notified with this: 
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/36df18de-c841-4c76-9d9a-ce7c5f9f81e4)
However, maybe I'm the only one, but if I get a bunch of source members open, then get pulled into a teams meeting, then go for a run, come back and start banging out some code changes, it would be nice to have the icon at the top (like in RDi) letting me know that I'm in browse mode.

**Comment by sebjulliand** [comment]
Hi!
I understand your issue here.

When a member is actually opened in read-only mode, it shows a lock (and doesn't let you type anything in the editor, too):
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/11096890/3a306110-6a6c-4485-9def-ae4410961ebe)

But that happens if the source member is open in read-only mode (which is something you do willingly).

In your case, the source member has been opened for edition but it's not possible to save it; which is something VSCode doesn't know until you try to actually save it.

One way to address this would be to check beforehand if that member can actually be written to.
@worksofliam, @chrjorgensen : I guess we can run `CHKOBJ OBJ(B/A) OBJTYPE(C) MBR(D) AUT(*WRITE)` before opening a source member and open in it in read-only mode if the command fails. What do you think?

**Reply by worksofliam** [reply]
@sebjulliand I think adding the `CHKOBJ` is a good idea, but I am concerned about adding more time to the member open code. 

I don't want to make it slower honestly, but then again that's the price we pay for good UX.

**Reply by sebjulliand** [reply]
Agreed. On second thought, maybe running `CHKOBJ *WRITE` on the source file is enough; I don't think you can restrict access to only a subset of members. In which case this could be tested when the source file gets expanded in the browser.

We'd have to run it when the member gets opened from the palette though.

As you said, a small price to pay for a good UX.

Let's see what @chrjorgensen thinks and I'll have a PR ready for that.

**Reply by Gregory1701** [reply]
I totally get (and appreciate) not wanting to slow down the open source member action.  What about a setting similar to this one:
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/df23543c-d5f1-4737-939d-e0b7ea09fcb0)
Where you can enter a comma delimited list of libraries to open read only all the time.  In my case, my production source library would be in here and any time I open a source member from a library in this new setting it's always read only.  How does that sound?

**Reply by chrjorgensen** [reply]
I think@sebjulliand's idea of checking authority when feeding the object browser is the best and quick solution. CHKOBJ is not expensive and could maybe be combined with the current read (to avoid a SSH roundtrip)?

We should also check when saving the member - authority could change / be changed while the member is open...

**Reply by chrjorgensen** [reply]
Server object authority must be the primary level - but having a list of libraries in the config for write protection as suggested by @Gregory1701 could be put on top. That would allow the user to be protected from editing these libraries by mistake.

Another thing: there should be a warning (message) presented to the user, when a member is opened as read-only due to lacking authority, since this is an unusual situation.

**Reply by Gregory1701** [reply]
Hey guys, any update on this issue/suggestion? 

@chrjorgensen - With regards to your comment "when a member is opened as read-only due to lacking authority, since this is an unusual situation" I disagree. I would argue (respectfully) that developers opening source members from their production library, which they do not have authority to directly change is the norm. Perhaps in a very small RPG shop, developers have edit access to the production library, but in my 29 years working in RPG shops, only my first job where I sat next to the president and vice-president of the small company did I have access to update production code. Given this, I feel a warning message every time you open source in display mode because you don't have access to it would be unpleasant. If a warning message is implemented, please include a parameter where we can turn it off.

Related to my original request, I would like to present for your consideration another scenario. Let's say, I'm working on a source member, so I have it in edit mode. I call a colleague to ask some questions. He/she does a Ctl+Alt+P and tries to open my source member so he can look at my code while we discuss it. (I think) since VS Code tries to open every source member for edit, it takes a while, but will eventually fail. In RDi, the random open of source members functionality has the ability to say you want to open in browse mode by default. This same option would be excellent if we could do the same in VS Code.

Thank you.

**Reply by sebjulliand** [reply]
Goodness! Good thing you digged up this one as I forgot about it 😅 
I'll have that PR done in the coming days. Let's recap:
- We'll run CHKOBJ at the source PF level to check the `*WRITE` authority
- We'll add a new connection setting which is a comma separated list of libraries that will be protected (same as having them in a protected filter)

This will be true for members opened from the browser and `CTRL+ALT+P` opened members.
Note: you can already open members in browse mode using `CTRL+ALT+O`.

@Gregory1701, @worksofliam , @chrjorgensen  : how does that sound to you?

**Reply by worksofliam** [reply]
@sebjulliand I think you're on to something for sure! Curious as to where this will be implemented. I am guessing in the QSYS FS provider?

**Reply by sebjulliand** [reply]
I didn't have a look at it yet but yes, probably something along those lines 😃

**Reply by chrjorgensen** [reply]
@Gregory1701 My remark about a message when a member is opened for readonly was written 6 months ago, but I still remember my thoughts back then: If the user does not have any indication of the source file or member being readonly (protected), there should be a message if the member is not editable and thus opened for read.

This message would almost never be seen - since the object browser will show the lock icon if the filter is protected or the user does not have *WRITE authority to the file. But the user could have a filter opened and another user locking the source file, by then no lock icon is shown for the first user and he/she would expect to be able to edit.

I agree fully on your arguments and your request for a default option for open for browse. This was my default back then in RDi, before I switched to C4i.

@sebjulliand Your recap looks fine - OK from me.

**Reply by sebjulliand** [reply]
I hear you guys 😄  I'll do something based on @Gregory1701's suggestion too: add an "Open" submenu to members that will offer to chose between "Edit" and "Browse". Left clicking on a member will still open it, but a new global setting will define if the member is to be opened in read-only mode by default or not. This should help addressing everything raised here.

`ctrl+alt+p` and `ctrl+alt+o` behavior will remain unchanged since both already offer to open in browse or edit mode.

**Reply by Gregory1701** [reply]
@worksofliam tagging you in per your request.
